src/rules/preferReadonlyRule.ts

```diff
@@ -0,0 +1,199 @@
+        `,
+        optionExamples: [true],
+        options: null,
+        optionsDescription: "Not configurable.",
+        requiresTypeInfo: true,
+        ruleName: "prefer-readonly",
+        type: "maintainability",
+        typescriptOnly: true,
+    };
+
+    public applyWithProgram(sourceFile: ts.SourceFile, program: ts.Program): Lint.RuleFailure[] {
+        const preferReadonlyWalker = new PreferReadonlyWalker(sourceFile, this.ruleName, program.getTypeChecker());
+        return this.applyWithWalker(preferReadonlyWalker);
+    }
+}
+
+class PreferReadonlyWalker extends Lint.AbstractWalker<void> {
+    private scope?: ClassScope;
+
+    public constructor(sourceFile: ts.SourceFile, ruleName: string, private readonly typeChecker: ts.TypeChecker) {
+        super(sourceFile, ruleName, undefined);
+    }
+
+    public walk(sourceFile: ts.SourceFile) {
+        if (sourceFile.isDeclarationFile) {
+            return;
+        }
+
+        ts.forEachChild(sourceFile, this.visitNode);
+    }
+
+    private readonly visitNode = (node: ts.Node) => {
+        if (utils.hasModifier(node.modifiers, ts.SyntaxKind.DeclareKeyword)) {
+            return;
+        }
+
+        switch (node.kind) {
+            case ts.SyntaxKind.ClassDeclaration:
+            case ts.SyntaxKind.ClassExpression:
+                this.handleClassDeclarationOrExpression(node as ts.ClassLikeDeclaration);
+                break;
+
+            case ts.SyntaxKind.Constructor:
+                this.handleConstructor(node as ts.ConstructorDeclaration);
+                break;
+
+            case ts.SyntaxKind.PropertyDeclaration:
+                this.handlePropertyDeclaration(node as ts.PropertyDeclaration);
+                break;
+
+            case ts.SyntaxKind.PropertyAccessExpression:
+                this.handlePropertyAccessExpression(node as ts.PropertyAccessExpression, node.parent!);
+                break;
+
+            default:
+                if (utils.isFunctionScopeBoundary(node)) {
+                    this.handleFunctionScopeBoundary(node);
+                } else {
+                    ts.forEachChild(node, this.visitNode);
+                }
+        }
+    }
+
+    private handleFunctionScopeBoundary(node: ts.Node) {
+        if (this.scope === undefined) {
+            ts.forEachChild(node, this.visitNode);
+            return;
+        }
+
+        this.scope.enterNonConstructorScope();
+        ts.forEachChild(node, this.visitNode);
+        this.scope.exitNonConstructorScope();
+    }
+
+    private handleClassDeclarationOrExpression(node: ts.ClassLikeDeclaration) {
+        const parentScope = this.scope;
+        const childScope = this.scope = new ClassScope(node, this.typeChecker);
+
+        ts.forEachChild(node, this.visitNode);
+
+        this.finalizeScope(childScope);
+        this.scope = parentScope;
+    }
+
+    private handleConstructor(node: ts.ConstructorDeclaration) {
+        if (this.scope !== undefined) {
```

Comment:
`this.scope` should never be undefined, because a constructor can only occur inside of a class